src/s3ql/backends/backblaze.py

```diff
@@ -0,0 +1,704 @@
+                # We probably can't use the connection anymore
+                self.conn.disconnect()
+            raise
+
+        self.sha1.update(buf)
+
+        # Check SHA1 on EOF
+        # (size == None implies EOF)
+        if (not buf or size is None) and not self.sha1_checked:
+            file_hash = self.resp.headers['x-bz-content-sha1'].strip('"')
+            self.sha1_checked = True
+            if file_hash != self.sha1.hexdigest():
+                log.warning('SHA mismatch for %s: %s vs %s',
+                            self.key, file_hash, self.sha1.hexdigest())
+                raise BadDigestError('BadDigest',
+                                     'SHA1 header does not agree with calculated SHA1')
+
+        return buf
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *a):
+        return False
+
+    def close(self, checksum_warning=True):
+        """Close object
+
+        If *checksum_warning* is true, this will generate a warning message if
+        the object has not been fully read (because in that case the MD5
+        checksum cannot be checked).
+        """
+
+        if self.closed:
+            return
+        self.closed = True
+
+        # If we have not read all the data, close the entire
+        # connection (otherwise we loose synchronization)
+        if not self.sha1_checked:
+            if checksum_warning:
+                log.warning("Object closed prematurely, can't check SHA1, and have to "
+                            "reset connection")
+            self.conn.disconnect()
+
+
+class ObjectW(object):
+    """An BackBlaze object open for writing"""
+
+    def __init__(self, key, backend, meta):
+        self.key = key
+        self.meta = meta
+        self.closed = False
+        self.obj_size = 0
+        self.backend = backend
+
+        # According to http://docs.python.org/3/library/functions.html#open
+        # the buffer size is typically ~8 kB. We process data in much
+        # larger chunks, so buffering would only hurt performance.
+        self.fh = tempfile.TemporaryFile(buffering=0)
+
+        self.sha1 = hashlib.sha1()
+
+    def write(self, buf):
+        self.fh.write(buf)
+        self.obj_size += len(buf)
+        self.sha1.update(buf)
+
+    def is_temp_failure(self, exc):
+        if isinstance(exc, NoSuchObject):
+            return True
+        return self.backend.is_temp_failure(exc)
+
+    @retry
+    def close(self):
+        """Close object and upload data"""
+        log.debug('started with %s', self.key)
+
+        if self.closed:
+            # still call fh.close, may have generated an error before
+            self.fh.close()
+            return
+
+        self.fh.seek(0)
+        upload_auth_token, upload_url = self.backend._get_upload_url()
+        upload_url = urllib.parse.urlparse(upload_url)
+
+        with HTTPConnection(upload_url.hostname, 443, ssl_context=self.backend.ssl_context) as conn_up:
```

Comment:
This means every upload will require creating a fresh connection (maybe even SSL). Is there a way to re-use the connection object? Hopefully the hostname of the upload url will not change every time..